packages/appium/docs/en/ecosystem/build-plugins.md

```diff
@@ -2,4 +2,436 @@
+out the [building drivers](./build-drivers.md) guide for further inspiration.)
+
+## Before you create your plugin
+
+Before creating your plugin, it's good to have a general idea of what you want your plugin to
+accomplish and whether it will be possible to implement it given the restrictions of the Appium
+platform. Reading through this guide will help you understand what's possible. In general, Appium's
+plugin system is extremely powerful and no attempts have been made to artificially limit what's
+possible with them (which is a main reason that all plugins are opt-in by the system administrator
+responsible for starting the Appium server---plugins are powerful and should only be used when
+explicitly trusted!).
+
+## Other plugins to reference
+
+There are a wide variety of open source Appium plugins available for perusal. It's definitely
+recommended to explore the code for some other plugins before embarking on writing your own. The
+Appium team maintains a set of official plugins in the [Appium GitHub
+repo](https://github.com/appium/appium). Links to other open source plugins can be found in the
+[Plugin list](./index.md#plugins)
+
+## Basic requirements for plugins
+
+These are the things your plugin *must* do (or be), if you want it to be a valid Appium plugin.
+
+### Node.js package with Appium extension metadata
+
+All Appium plugins are fundamentally Node.js packages, and therefore must have a valid
+`package.json`. Your driver is not _limited_ to Node.js, but it must provide an adapter written in
+Node.js so it can be loaded by Appium.
+
+Your `package.json` must include `appium` as a `peerDependency`. The requirements for the
+dependency versions should be as loose as possible (unless you happen to know your plugin will only
+work with certain versions of Appium). For Appium 2.0, for example, this would look something like
+`^2.0.0`, declaring that your plugin works with any version of Appium that starts with 2.x.
+
+Your `package.json` must contain an `appium` field, like this (we call this the 'Appium extension
+metadata'):
+
+    ```json
+    {
+      ...,
+      "appium": {
+        "pluginName": "fake",
+        "mainClass": "FakePlugin"
+      },
+      ...
+    }
+    ```
+
+The required subfields are:
+
+* `pluginName`: this should be a short name for your plugin.
+* `mainClass`: this is a named export (in CommonJS style) from your `main` field. It must be a
+  class which extends Appium's `BasePlugin` (see below).
+
+### Extend Appium's `BasePlugin` class
+
+Ultimately, your plugin is much easier to write because most of the hard work of defining patterns
+for overriding commands is done for you. This is
+all encoded up as a class which Appium exports for you to use, called `BasePlugin`. It is exported
+from `appium/plugin`, so you can use one of these styles to import it and create your *own* class
+that extends it:
+
+```js
+import {BasePlugin} from 'appium/plugin';
```

Comment:
I assume we should also mention the base plugin must be imported as a peer dependency for this to work?

Reply:
it's appium that must be imported as a peer dependency, which is already mentioned above